backend/src/main/java/de/bund/digitalservice/ris/norms/application/service/XmlDocumentExtractor.java

```diff
@@ -0,0 +1,167 @@
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import org.springframework.stereotype.Service;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+
+/**
+ * Service class within the application core part of the backend. It is responsible for working with
+ * w3c documents. It is annotated with {@link Service} to indicate that it's a service component in
+ * the Spring context.
+ */
+@Service
+public class XmlDocumentExtractor {
+
+  /**
+   * Create a deep copy of a Document
+   *
+   * @param originalDocument the document we want to clone
+   * @return a deep copy of the <code>originalDocument</code>
+   */
+  Document cloneDocument(Document originalDocument) {
+    try {
+      Node originalRootNode = originalDocument.getDocumentElement();
+      DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+      DocumentBuilder db = dbf.newDocumentBuilder();
+
+      Document clonedDocument = db.newDocument();
+      Node clonedRootNode = clonedDocument.importNode(originalRootNode, true);
+      clonedDocument.appendChild(clonedRootNode);
+      return clonedDocument;
+    } catch (ParserConfigurationException e) {
+      throw new XmlProcessingException();
+    }
+  }
+
+  ReplacementPair getReplacementText(Node firstModificationNodeInAmendingLaw) {
+    try {
+      final Optional<String> oldText = getTextToBeReplaced(firstModificationNodeInAmendingLaw);
+
+      final Optional<String> newText = getNewTextInReplacement(firstModificationNodeInAmendingLaw);
+      if (oldText.isEmpty() || newText.isEmpty()) throw new ModificationException();
+      return new ReplacementPair(oldText.get(), newText.get());
+    } catch (XPathExpressionException e) {
+      throw new XmlProcessingException();
+    }
+  }
+
+  record ReplacementPair(String oldText, String newText) {}
+
+  Node findTargetLawNodeToBeModified(Document targetLaw, Node firstModificationNodeInAmendingLaw) {
+    try {
+      final Optional<String> modificationHref;
+
+      modificationHref = findHrefInModificationNode(firstModificationNodeInAmendingLaw);
+      if (modificationHref.isEmpty()) throw new ModificationException();
+
+      final Optional<String> eId = getEIdfromModificationHref(modificationHref.get());
+      if (eId.isEmpty()) throw new ModificationException();
+
+      final Optional<Node> targetLawNodeToBeModified = findNodeByEId(eId.get(), targetLaw);
+      if (targetLawNodeToBeModified.isEmpty()) throw new ModificationException();
+      return targetLawNodeToBeModified.get();
+    } catch (XPathExpressionException e) {
+      throw new XmlProcessingException();
+    }
+  }
+
+  /**
+   * Get the first modification in the amending law that can be found
+   *
+   * @param amendingLaw the law to be searched
+   * @return the Node containing the modification
+   */
+  Node getFirstModification(Document amendingLaw) {
```

Comment:
I wonder about the place of these functions. They are too specific for a general `XmlHelper` as they rely on the semantics of the TimeMachine.